### PR TITLE
Add OS Timer driver for MXRT600

### PR DESCRIPTION
**Pull request by mmahadevan108**
On MXRT600, switch the kernel timer to OS timer block available on MXRT600 from SYSTICK timer.

### CODE REVIEW

File: dts/arm/nxp/nxp_rt6xx_common.dtsi
Comment (galak) [inline dts/arm/nxp/nxp_rt6xx_common.dtsi:45]: If this is the case, why is os_timer status `disabled` by default?

File: arch/arm/core/aarch32/cortex_m/timing.c
Comment (galak) [inline arch/arm/core/aarch32/cortex_m/timing.c:32]: Should we really be changing IMX and KINETIS?

File: west.yml
Comment (andyross) [inline west.yml:103]: I'll let someone else decide on policy, but it can't possibly be kosher to have a in-tree module pointed at a live pull request and not a committed tree.

**Review by andyross — CHANGES_REQUESTED** [review]
One bug with tick clamping, a design point about device tree, various notes

File: soc/arm/nxp_imx/rt6xx/Kconfig.soc
Comment (MaureenHelm) [inline soc/arm/nxp_imx/rt6xx/Kconfig.soc:17]: Don't remove this, without it we can't ever use SYSTICK

File: soc/arm/nxp_imx/rt6xx/Kconfig.defconfig.mimxrt685_cm33
Comment (MaureenHelm) [inline soc/arm/nxp_imx/rt6xx/Kconfig.defconfig.mimxrt685_cm33:43]: wrap with `if MCUX_OS_TIMER..endif`

File: dts/arm/nxp/nxp_rt6xx_common.dtsi
Comment (MaureenHelm) [reply dts/arm/nxp/nxp_rt6xx_common.dtsi:48]: How is this lying about the hardware? The hardware is still there, it's just not enabled. Nordic does the same thing: https://github.com/zephyrproject-rtos/zephyr/blob/c59cf6008b605523ba772ea8d29279dac2bf90b6/dts/arm/nordic/nrf_common.dtsi#L19-L25

I agree we should be able to switch between the SYSTICK and the MCUX OS Timer driver

File: west.yml
Comment (MaureenHelm) [reply west.yml:103]: It's not. This is an intermediate step that allows us to verify both the HAL PR and the main zephyr repo PR in CI. Once the HAL PR gets merged, this will get updated to the new SHA before we merge this PR.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: dts/arm/nxp/nxp_rt6xx_common.dtsi
Comment (mmahadevan108) [reply dts/arm/nxp/nxp_rt6xx_common.dtsi:45]: Thanks for catching this, I have fixed it.

File: arch/arm/core/aarch32/cortex_m/timing.c
Comment (mmahadevan108) [reply arch/arm/core/aarch32/cortex_m/timing.c:32]: IMX was a mistake. I have changed this to RT family. I have added NXP families in anticipation future power management changes.

File: soc/arm/nxp_imx/rt6xx/Kconfig.soc
Comment (mmahadevan108) [reply soc/arm/nxp_imx/rt6xx/Kconfig.soc:17]: Sure, thank you for spotting this.

File: drivers/timer/mcux_os_timer.c
Comment (mmahadevan108) [reply drivers/timer/mcux_os_timer.c:28]: Done

File: drivers/timer/Kconfig
Comment (mmahadevan108) [reply drivers/timer/Kconfig:355]: I have changed to be conditional if the OS Timer is enabled.

File: drivers/timer/mcux_os_timer.c
Comment (mmahadevan108) [reply drivers/timer/mcux_os_timer.c:91]: This was added in error, I have deleted these lines

File: dts/arm/nxp/nxp_rt6xx_common.dtsi
Comment (galak) [reply dts/arm/nxp/nxp_rt6xx_common.dtsi:48]: exactly, we might want to default both to disabled in the soc dtsi files and have the board than set which one it wants.

File: drivers/timer/Kconfig
Comment (galak) [reply drivers/timer/Kconfig:355]: I wonder if we default this based on the devicetree status?

File: soc/arm/nxp_imx/rt6xx/Kconfig.defconfig.mimxrt685_cm33
Comment (mmahadevan108) [reply soc/arm/nxp_imx/rt6xx/Kconfig.defconfig.mimxrt685_cm33:45]: Done

File: dts/arm/nxp/nxp_rt6xx_common.dtsi
Comment (mmahadevan108) [reply dts/arm/nxp/nxp_rt6xx_common.dtsi:48]: Thanks for the suggestion. I have disabled both timers by default and enabled os_timer in the board dtsi file. I have tested both scenarios with os_tick and systick.

File: arch/arm/core/aarch32/cortex_m/timing.c
Comment (mmahadevan108) [reply arch/arm/core/aarch32/cortex_m/timing.c]: I have updated this code to make this change for RT600 only. We can add changes for LPC and other RT devices later whenever it is required.

File: dts/arm/nxp/nxp_rt6xx_common.dtsi
Comment (MaureenHelm) [reply dts/arm/nxp/nxp_rt6xx_common.dtsi:45]: Since you disable systick at the SoC level, you should also enable os_timer at the SoC level. Currently it's done at the board level

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
@andyross please revisit

File: dts/arm/nxp/nxp_rt6xx_common.dtsi
Comment (mmahadevan108) [reply dts/arm/nxp/nxp_rt6xx_common.dtsi:45]: It gives each board the flexibility to pick, do we not want this?

Comment (MaureenHelm) [reply dts/arm/nxp/nxp_rt6xx_common.dtsi:45]: Each board will still have the flexibility to pick using a dts overlay.

**Comment by MaureenHelm** [comment]
Please rebase

**Comment by mmahadevan108** [comment]
> Please rebase

Rebase done.

**Review by MaureenHelm — APPROVED** [review]

**Comment by mmahadevan108** [comment]
Anything else needed for this PR?

**Comment by mmahadevan108** [comment]
@galak @andyross could you please if anything is needed for this PR?

**Comment by mmahadevan108** [comment]
Anything else needed for this PR.

**Review by andyross — APPROVED** [review]
Timer driver looks good.  As far as the DTS changes, it still seems wrong to be playing with a "disabled" flag in *hardware* configuration when the hardware (for both timers) works fine in all circumstances.  As I understand the architecture, the choice of timer driver is a 100% software decision and belongs in kconfig, no?

**Review by galak — APPROVED** [review]

**Comment by mmahadevan108** [comment]
> Timer driver looks good. As far as the DTS changes, it still seems wrong to be playing with a "disabled" flag in _hardware_ configuration when the hardware (for both timers) works fine in all circumstances. As I understand the architecture, the choice of timer driver is a 100% software decision and belongs in kconfig, no?

**Pull request merged**
